File: CCC03S3.cpp
// Problem: CCC '03 S3 - Floor Plan
// Contest: DMOJ
// URL: https://dmoj.ca/problem/ccc03s3
// Memory Limit: 64 MB
// Time Limit: 1000 ms
// Ryan Liu

#include <bits/stdc++.h>
using namespace std;

#define pb push_back

int m, w, l, x, y, tmr, numr = 0;

vector<int> roomarea;
queue<int> q;
bool used[25][25];
char mc[25][25];
string t;

void bfs(int b, int a) {
  tmr = 1;
  q.push(a);
  q.push(b);
  used[b][a] = true;
  while (!q.empty()) {
    int v1 = q.front();
    q.pop();
    int v2 = q.front();
    q.pop();

    /*for (int u : adj[v]) {
      if (!used[u]) {
        used[u] = true;
        q.push(u);
        dist[u] = dist[v] + 1;
      }

    }*/
    if ((v2 - 1 >= 0) && (!used[v2 - 1][v1]) && mc[v2 - 1][v1] == '.') {
      used[v2 - 1][v1] = true;
      q.push(v1);
      q.push(v2 - 1);
      ++tmr;
    }
    if ((v2 + 1 < l) && (!used[v2 + 1][v1]) && mc[v2 + 1][v1] == '.') {
      used[v2 + 1][v1] = true;
      q.push(v1);
      q.push(v2 + 1);
      ++tmr;
    }
    if ((v1 - 1 >= 0) && (!used[v2][v1 - 1]) && mc[v2][v1 - 1] == '.') {
      used[v2][v1 - 1] = true;
      q.push(v1 - 1);
      q.push(v2);
      ++tmr;
    }
    if ((v1 + 1 < w) && (!used[v2][v1 + 1]) && mc[v2][v1 + 1] == '.') {
      used[v2][v1 + 1] = true;
      q.push(v1 + 1);
      q.push(v2);
      ++tmr;
    }
  }
  roomarea.pb(tmr);
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);
  cout.tie(0);
  memset(used, false, sizeof(used));

  cin >> m >> l >> w;

  // get input
  for (int i = 0; i < l; i++) {
    for (int j = 0; j < w; j++) {
      cin >> mc[i][j];
    }
  }

  for (int i = 0; i < l; i++) {
    for (int j = 0; j < w; j++) {
      if (mc[i][j] == '.' && !used[i][j]) {
        bfs(i, j);
      }
    }
  }

  sort(roomarea.rbegin(), roomarea.rend());

  for (int i = 0; i < roomarea.size(); i++) {
    if ((m - roomarea[i]) >= 0) {
      m -= roomarea[i];
      numr++;

    } else {
      break;
    }
  }

  if (numr == 1) {
    cout << 1 << " room, " << m << " square metre(s) left over"
         << "\n";
    return 0;
  }

  cout << numr << " rooms, " << m << " square metre(s) left over"
       << "\n";
}

File: CCC03S3.java
import java.io.*;
import java.util.*;
/**
 * CCC '03 S3 - Floor Plan
 * Question type: Graph Theory
 * 100/100 on DMOJ
 * Question URL: https://dmoj.ca/problem/ccc03s3
 * @author Tommy Pang
 */

public class CCC03S3 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
    static StringTokenizer st;
    static int R, C, F, mod = (int) 1e9+7;
    static char[][] floor;
    static List<Integer> rooms = new ArrayList<>();
    static int [] dr = new int[] {0, 0, -1, 1}, dc = new int[] {-1, 1, 0, 0};
    static boolean [][] vis;
    public static void main(String[] args) throws IOException {
        F = readInt(); R = readInt(); C = readInt();
        floor = new char[R][C]; vis = new boolean[R][C];
        for (int i = 0; i < R; i++) {
            floor[i] = readLine().toCharArray();
        }
        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (floor[i][j]=='.' && !vis[i][j]) {
                    rooms.add(BFS(i, j));
                }
            }
        }
        rooms.sort(Collections.reverseOrder());
        for (int i = 0; i < rooms.size(); i++) {
            if (F>=rooms.get(i)) {
                F-=rooms.get(i);
            }
            else {
                if (i==1) {
                    System.out.println(i + " room, " + F + " square metre(s) left over");
                }
                else System.out.println(i + " rooms, " + F + " square metre(s) left over");
                return;
            }
        }
        System.out.println(rooms.size() + " rooms, " + F + " square metre(s) left over");
    }
    public static int BFS(int r, int c) {
        int cnt = 1;
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{r, c});
        vis[r][c] = true;
        while (!queue.isEmpty()) {
            int [] cur = queue.poll();
            int i = cur[0], j = cur[1];
            for (int k = 0; k < 4; k++) {
                if (valid(i+dr[k], j+dc[k])) {
                    queue.add(new int[] {i+dr[k], j+dc[k]});
                    vis[i+dr[k]][j+dc[k]] = true;
                    cnt++;
                }
            }
        }
        return cnt;
    }
    public static boolean valid(int i, int j) {
        if (i<0 || i==R || j<0 || j==C || vis[i][j] || floor[i][j]=='I') return false;
        return true;
    }

    static String next() throws IOException {
        while (st == null || !st.hasMoreTokens())
            st = new StringTokenizer(br.readLine().trim());
        return st.nextToken();
    }
    static long readLong() throws IOException {
        return Long.parseLong(next());
    }
    static int readInt() throws IOException {
        return Integer.parseInt(next());
    }
    static double readDouble() throws IOException {
        return Double.parseDouble(next());
    }
    static char readCharacter() throws IOException {
        return next().charAt(0);
    }
    static String readLine() throws IOException {
        return br.readLine().trim();
    }
}
